### PR TITLE
Fix two different version of a spockframework on the classpath

### PR DESCRIPTION
**Pull request by fvclaus**
I was curious to see how Buildship works by running a few tests, but I had a hard time getting the test to run in Eclipse. When I imported the project from the Oomph catalog and ran the core tests, this error occurred:
```
org.spockframework.runtime.extension.ExtensionException: Duplicated Extension declaration for [org.spockframework.runtime.extension.builtin.IncludeExcludeExtension]
Source 1: bundleresource://49.fwk811369097:36/META-INF/services/org.spockframework.runtime.extension.IGlobalExtension
Source 2: bundleresource://49.fwk811369097:20/META-INF/services/org.spockframework.runtime.extension.IGlobalExtension
This is most likely caused by having two different version of a library on the classpath.
	at org.spockframework.runtime.ExtensionClassesLoader.loadClasses(ExtensionClassesLoader.java:40)
	at org.spockframework.runtime.ExtensionClassesLoader.loadClassesFromDefaultLocation(ExtensionClassesLoader.java:31)
	at org.spockframework.runtime.RunContext.createBottomContext(RunContext.java:185)
	at org.spockframework.runtime.RunContext.get(RunContext.java:153)
	at org.spockframework.runtime.Sputnik.runExtensionsIfNecessary(Sputnik.java:90)
	at org.spockframework.runtime.Sputnik.getDescription(Sputnik.java:54)
	at org.junit.runners.model.RunnerBuilder.configureRunner(RunnerBuilder.java:81)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:72)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.requests.ClassRequest.createRunner(ClassRequest.java:28)
	at org.junit.internal.requests.MemoizingRequest.getRunner(MemoizingRequest.java:19)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.createUnfilteredTest(JUnit4TestLoader.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.createTest(JUnit4TestLoader.java:76)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestLoader.loadTests(JUnit4TestLoader.java:49)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:513)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:74)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

After a lot of debugging, I found out that two spock library are on the classpath of the `core` plugin. One from `core.tests` and one from `oomph.tests`. They are both loaded, because the test configuration for the tests specify to launch with `all workspace and enabled target plug-ins`. It can be deselected, but this blows up the size of the test configuration and I don't know how maintainable this is. Instead, I introduced a `testdependency` project. This also removes the duplication of the library jar files.

Next, the test configuration specifies Java 8 as execution environment. This must be a mistake. All plugins require Java 11. 
Last, the tests require the `-Djdk8.location` and `-Djdk11.location` to be set. 

With these changes, the test configurations run in Eclipse and on the command-line with Gradle.

I tried to setup the variables for the test configuration with Oomph, but I couldn't verify whether or not this actually works, because I wasn't able to redirect my Oomph setup to use my local copy. I tried for at least 8 hours, digging around the source code, but finally gave up. The configuration under `oomph-setup/README.md` had no effect for me (`master` version from Github is used). Maybe I am doing something wrong. If somebody can confirm if `oomph-setup/README.md` is still up-to-date, I will give it another try.

I hope not all of my changes are complete nonsense.

### CODE REVIEW

**Comment by donat** [comment]
Thanks for looking into this. Those items were on my list, but since I thought nobody is actively looking into the code they had lower priority. I'll test your changes soon.

**Comment by FieteO** [comment]
> Thanks for looking into this. Those items were on my list, but since I thought nobody is actively looking into the code they had lower priority. I'll test your changes soon.

That everyone is having such a hard time setting up this project may be a bigger reason why no one is looking into the code. I also have the ambition to contribute something but can't, since I am unable to get this running...

**Comment by donat** [comment]
Thanks for not giving up. I hope the project dev environment will be in better shape soon.

**Review by donat — APPROVED** [review]
I've tested the IDE test execution as well as executing the `eclipseTest` on the command line and your changes work well! Once again, thank you for investing time into this. 

I've identified a couple of things that should be looked at:
- When importing the project into Eclipse via the Oomph setup model, the testdependencies subproject does not appear.
- There's a way to regenerate the jars and the project metadata for test dependencies which may need a change. 
- The Gradle build should have an actionable error message when a required property is not present. Instead of ` jdk8.location project property required for build` it should say:
```
Required property 'jdk8.location' is not defined. Consider adding all required elements to gradle properties file (default location: ~/.gradle/gradle.properties): 

# required properties for the Buildship build
jdk8.location=/path/to/jdk-8
jdk11.location=/path/to/jdk-11
``` 

With that said, the PR is fine and I can merge it it as-is. Let me know if you want to work on the items above, otherwise I'll eventually do them.

**Comment by donat** [comment]
thank you @fvclaus!

File: org.eclipse.buildship.ui.test/Launch Buildship UI tests.launch [hunk 1]
```diff
@@ -27,13 +27,14 @@
...
     <booleanAttribute key="org.eclipse.jdt.launching.ATTR_USE_START_ON_FIRST_THREAD" value="true"/>
-    <stringAttribute key="org.eclipse.jdt.launching.JRE_CONTAINER" value="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+    <stringAttribute key="org.eclipse.jdt.launching.JRE_CONTAINER" value="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11"/>
     <stringAttribute key="org.eclipse.jdt.launching.MAIN_TYPE" value=""/>
     <stringAttribute key="org.eclipse.jdt.launching.PROGRAM_ARGUMENTS" value="-os ${target.os} -ws ${target.ws} -arch ${target.arch} -nl ${target.nl} -consoleLog"/>
     <stringAttribute key="org.eclipse.jdt.launching.PROJECT_ATTR" value="org.eclipse.buildship.ui.test"/>
     <stringAttribute key="org.eclipse.jdt.launching.SOURCE_PATH_PROVIDER" value="org.eclipse.pde.ui.workbenchClasspathProvider"/>
-    <stringAttribute key="org.eclipse.jdt.launching.VM_ARGUMENTS" value="-ea"/>
+    <stringAttribute key="org.eclipse.jdt.launching.VM_ARGUMENTS" value="-ea -Djdk8.location=${jre.location-1.8} -Djdk11.location=${jre.location-11}"/>
```
[the remark this answers is not in the document]
Comment (fvclaus) [reply org.eclipse.buildship.ui.test/Launch Buildship UI tests.launch:37]: Adressed in #1156

**Comment by fvclaus** [comment]
Also: Thanks for patching my MR. You really did not give me any time at all to adress those issues :)

**Comment by donat** [comment]
If you look at `buildSrc` there's a `BundlePlugin` class that defines the `updateLibs` task. The intent of the task is to copy the dependencies to the `lib` folder and add them to the classpath. I'm not sure if it still works; I'll leave it to you to find out.

**Pull request merged**